hypothesis/lms
# Support non-LTI authenticated launches of the VitalSource book viewer

Add a new method of launching the VitalSource Bookshelf reader in VS assignments, which does not use an LTI launch. This change is gated behind the `vitalsource_anon_launch` feature flag:

- When the `vitalsource_anon_launch` flag is disabled, VS assignment launches will launch the Bookshelf reader app using an LTI launch, i.e. similar to how LMSes launch Hypothesis. This passes information about the LTI user through to VitalSource.

- When the `vitalsource_anon_launch` flag is enabled, VS assignment launches load the final `https://hypothesis.vitalsource.com/books/{book_id}/cfi/{cfi}` URL directly in an iframe. The user will need to login using their VitalSource credentials before they can start to access the assignment.

The implementation works by having the `VitalSourceService` factory check the status of the flag and pass the LTI launch key/secret to `VitalSourceService` if LTI launches are enabled or omit it otherwise. When an assignment launch configuration is generated, the `vitalSource.{launchUrl, launchParams}` properties change depending on whether the service has an LTI launch key/secret. If it does, the `launchParams` are set and the frontend creates and submits a hidden form. If no LTI launch key is set, the `launchParams` are omitted and the frontend loads the `launchUrl` directly in an iframe.

This new non-authenticated launch method is expected to be a temporary step towards some other authentication method in future. The details of that are being worked out. The reason for dropping the LTI launch is because it has unintended side effects (see  https://github.com/hypothesis/private-issues/issues/102) that we need to avoid.

Part of https://github.com/hypothesis/private-issues/issues/102

---

**Testing:**

1. Set the following environment variables before running `make dev`:

   ```sh
   export FEATURE_FLAGS_ALLOWED_IN_COOKIE=vitalsource_anon_launch
   export VITALSOURCE_API_KEY={See 1Password}
   export VITALSOURCE_LTI_LAUNCH_KEY={See 1Password}
   export VITALSOURCE_LTI_LAUNCH_SECRET={See 1Password}
   ```

2. Create a new VitalSource assignment (example book URL: https://bookshelf.vitalsource.com/reader/books/9781400847402/epubcfi/6/24[%3Bvnd.vst.idref%3Dch3]!/4)
3. Launch the assignment with the `vitalsource_anon_launch` flag disabled. The VS viewer should load. If you inspect the document loads that happen as the assignment launches, you'll see that a form is submitted to `https://bc.vitalsource.com` with LTI user information proxied from the LTI launch
4. Visit http://localhost:8001/flags and enable the `vitalsource_anon_launch` flag
5. Re-launch the assignment and inspect document loads in devtools. This time the `https://hypothesis.vitalsource.com/...` viewer URL loads directly via a GET request
